constraints/miden-vm/decoder.air

```diff
@@ -0,0 +1,483 @@
+    # flag for end, repeat, respan, halt
+    let f_errh = b[6] * b[5] * b[4]
+    return f_sjsl + f_errh + f_call(b, extra) + f_syscall(b, extra)
+
+
+fn f_ctrli(b: vector[7], extra: scalar) -< scalar:
+    return f_join(b) + f_split(b) + f_loop(b) + f_call(b, extra) + f_syscall(b, extra)
+
+
+# Returns transition label, composed of the operation label and the periodic columns that uniquely 
+# identify each transition function.
+fn get_transition_label(op_label: scalar) -> scalar:
+    return op_label + 2^4 * cycle_row_7 + 2^5 * cycle_row_0
+
+
+fn get_f_g8(op_batch_flags: vector[3]) -> scalar:
+    return op_batch_flags[0]
+
+
+fn get_f_g4(op_batch_flags: vector[3]) -> scalar:
+    return binary_not(op_batch_flags[0]) * op_batch_flags[1] * op_batch_flags[2]
+
+
+fn get_f_g2(op_batch_flags: vector[3]) -> scalar:
+    return binary_not(op_batch_flags[0]) * binary_not(op_batch_flags[1]) * op_batch_flags[2]
+
+
+fn get_f_g1(op_batch_flags: vector[3]) -> scalar:
+    return binary_not(op_batch_flags[0]) * op_batch_flags[1] * binary_not(op_batch_flags[2])
+
+
+### Helper evaluators #############################################################################
+
+# Enforces that column must be binary.
+ev is_binary(main: [a]):
+    enf a^2 = a
+
+
+# Enforces that value in column is copied over to the next row.
+ev is_unchanged(main: [column]):
+    ev column' = column
+
+
+# Enforces decoder general constraints.
+ev general_constraints(main: [addr, op_bits[7], hasher[8], is_span, s0], aux: [extra]):
+    # TODO: add constraints to check not used operational bits.
+
+    # Enforce that when SPLIT or LOOP operation is executed, the top of the operand stack must 
+    # contain a binary value.
+    enf is_binary([s0]) when f_split(op_bits) | f_loop(op_bits)
+
+    # Enforce that When REPEAT operation is executed, the value at the top of the operand stack 
+    # must be 1.
+    enf s0 = 1 when f_repeat(op_bits, extra)
+
+    # Enforce that when REPEAT operation is executed, the value in hasher[4] column (the 
+    # is_loop_body flag), must be set to 1. This ensures that REPEAT operation can be executed only
+    # inside a loop.
+    enf hasher[4] = 1 when f_repeat(op_bits, extra)
+
+    # Enforce that when RESPAN operation is executed, we need to make sure that the block ID is 
+    # incremented by 8.
+    enf addr' = addr + 8 when f_respan(op_bits, extra)
+
+    # Enforce that when END operation is executed and we are exiting a loop block (i.e., is_loop, 
+    # value which is stored in hasher[5], is 1), the value at the top of the operand stack must be 
+    # 0.
+    enf s0 = 0 when f_end(op_bits, extra) & hasher[5]
+
+    # Enforce that when END operation is executed and the next operation is REPEAT, values in 
+    # hasher[0], ..., hasher[4] (the hash of the current block and the is_loop_body flag) must be 
+    # copied to the next row.
+    enf is_unchanged([hasher[i]]) when f_end(op_bits, extra) & f_repeat_next(op_bits, extra) for i in 0..5
```

Comment:
I'd like to change the order on this. I think it's much more readable to do the following:

```
enf is_unchanged([hasher[i]]) for i in 0..5 when f_end(op_bits, extra) & f_repeat_next(op_bits, extra)
```

I've already mentioned it to @tohrnii in the constraint comprehension PR, and this update is planned for the future.